lib/translator/inquiry.cc

```diff
@@ -13,10 +13,104 @@
+  // check if nonzero
+  bool nguid_nz =
+      identify_namespace_data.nguid[0] || identify_namespace_data.nguid[1];
+  bool eui64_nz = identify_namespace_data.eui64;
+  uint8_t identifier_length;
+
+  if (nguid_nz) {
+    // Shall be set to 08h if the EUI-64 field is populated from the NVMe EUI64
+    // field.
+    identifier_length = kIdentifierLengthNguid;
+
+    uint8_t identification_descriptor_identifier[16];
+
+    memcpy(&identification_descriptor_identifier,
+           &identify_namespace_data.nguid,
+           sizeof(identification_descriptor_identifier));
+
+  } else if (eui64_nz) {
+    // Shall be set to 10h if the EUI-64 field is populated from the NVMe NGUID
+    // field.
+    identifier_length = kIdentifierLengthEui64;
+
+    uint64_t identification_descriptor_identifier =
+        identify_namespace_data.eui64;
+  } else {
+    // function not supported as both NGUID and EUI64 fields are zero
+    DebugLog("Both NGUID and EUI-64 fields are zero in IdentityNamespaceData");
+  }
+
+  scsi_defs::IdentificationDescriptor identification_descriptor = {
+
+      // Shall be set to 0h. PIV field shall indicate this field is reserved as
+      // no
+      // specific protocol to be identified. This field will be ignored!
+      .protocol_identifier = scsi_defs::ProtocolIdentifier::kFibreChannel,
+
+      // Shall be set to 1h indicating associated fields are in binary format.
+      .code_set = scsi_defs::CodeSet::kBinary,
+
+      // Shall be set to 0b indicating PROTOCOL IDENTIFIER field is reserved.
+      .protocol_identifier_valid = 0,
+
+      // Shall be set to 00b indicating DESIGNATOR field is associated with
+      // logical unit.
+      .association = scsi_defs::Association::kPhysicalDevice,
+
+      // Shall be set to 2h indicating EUI-64 based identifier.
+      .identifier_type = scsi_defs::IdentifierType::kEUI64,
+
+      .identifier_length = identifier_length};
+  return identification_descriptor;
+}
+}  // namespace
+
+// Creates and validates a Inquiry Command struct
+StatusCode RawToScsiCommand(absl::Span<const uint8_t> raw_cmd,
+                            scsi_defs::InquiryCommand &cmd) {
+  if (raw_cmd.empty()) {
+    DebugLog("buffer is empty or nullptr");
+    return StatusCode::kInvalidInput;
+  }
+
+  uint8_t raw_opcode = raw_cmd[0];
+  if (raw_opcode > 0xaf) {
+    DebugLog("raw opcode is out of range: %x", raw_opcode);
+    return StatusCode::kInvalidInput;
+  }
+
+  scsi_defs::OpCode opcode = static_cast<scsi_defs::OpCode>(raw_opcode);
+
+  if (opcode != scsi_defs::OpCode::kInquiry) {
+    absl::string_view expected_cmd_str =
+        ScsiOpcodeToString(scsi_defs::OpCode::kInquiry);
+    absl::string_view cmd_str = ScsiOpcodeToString(opcode);
+    DebugLog("wrong opcode. expected %s got %s.", expected_cmd_str, cmd_str);
+    return StatusCode::kInvalidInput;
+  }
+
+  // TODO: check invalid parameters in scsi_command
+  cmd = *reinterpret_cast<const scsi_defs::InquiryCommand *>(&raw_cmd[1]);
```

Comment:
Should use ReadValue https://github.com/googleinterns/scsi2nvme/blob/master/lib/translator/common.h#L32